README.md
# phpTradfri
This project is a set of PHP Classes working with [libcoap](https://github.com/obgm/libcoap)'s coap-client to allow controlling IKEA's Trådfri smart lighting products. For this, coap-client is used to communicate with the Trådfri Gateway through [COAP](https://tools.ietf.org/html/rfc7252) over TLS.
## Features
* Querying a list of the following items configured in the gateway:
  * Devices
    * Lamps
    * Remote Controls
    * Motion Sensors
  * Groups
* Querying the status of the items 
  * On/Off
  * Brightness
  * Battery level
  * Names
* Turning lights and groups on and off
* Gateway Functions
  * Querying the Status informations
  * Reboot
## Requirements
* IKEA Trådfri Gateway
* PHP7 (with minor adjustments older versions can also be used)
* libcoap (only supports Linux unfortunately)
## Installation
1. Install coap-client with DTLS support. Preferrably, debug output from [tinydtls](https://projects.eclipse.org/projects/iot.tinydtls) should be disabled (otherwise adjustments in the scripts are required, see inside list.php for details on this). The script [install-coap-client.sh](https://github.com/ggravlingen/pytradfri/blob/master/script/install-coap-client.sh) from the [pytradfri](https://github.com/ggravlingen/pytradfri) repository automates this.
2. Place the scripts in a web server directory accessible to web browsers. Ensure executing processes on the command line is allowed in your PHP installation.
## Configuration
While creating an object in your code, you must provide the parameters for user, secretkey and IP Address
Example:
```
$groups = new tradfrigroups("<user>", "<secret>", "<ip>");
```


Release-Notes.md
# phpTradfri - Version History
## Version 2.4 - 25th Sep 2019
* groups.php
  + add getMembers() Function
## Version 2.3 - 22nd Sep 2019
* devices.php
  + add Last Seen Information for Lamps and Remote Controls
  + add statuscontroloutlet()
  + add functions for dimming lamps
  + extend poweron() and poweroff() for control outlets
  + extend getPowerStatus() for control outlets
* gateway.php
  * bugfixing ntp in status
* general.php
  + adding return for action()
* groups.php
  + add functions for dimming lamps
## Version 2.2 - 16th Sep 2019
+ getDimmer() and setDimmer() added to devices.php
## Version 2.1.1 - 14th Sep 2019
* Fix missing return in Function statusgateway()
## Version 2.1 - 14th Sep 2019
* Add Subclass for Trådfri Gateway
* defines.php
  + defines for gateway
* general.php
  * extend function action() for Gateway Class (reboot command has no payload to transmit)
* include.php
  + add gateway.php for including gateway Sublcass
* ReadMe
  + Gateway functions
* Release Notes
  * create Release Notes
## Version 2.0 - 11th Sep 2019
* general.php
  * Move from Constants to Parameters for Gateway Config
  * to initialize new object parameters User, Secret and Gateway IP must enterd. Example:
    ```
    $groups = new tradfrigroups("<user>", "<secret>", "<ip>");
    ```
  * cleaning up file
* Update ReadMe File
## Version 1.1 - 09th Sep 2019
* devices
  + Output of Firmware Version at statusremotecontrol()
  + Add Motion Sensor to statusremotecontrol()
## Version 1.0 - 30th Aug 2019
* Initial Commit

groups.php
<?php

require_once('general.php');

class tradfrigroups extends tradfri
	{

	function getDimmer($Id){

		$dimid = $this->getDetails("15004/$Id");

		return $dimid[DIMMER];

		}

	function getIds(){

		return explode(",", trim(str_replace(['[',']'], "" ,strstr($this->query("15004"), '[13'))));

		}

	function getIdbyName($name){

		$Ids = $this->getIds() ?? [];
		foreach($Ids as $Id){
			$Idname = strtolower(trim($this->getName("15004/$Id")));
			$Idname = str_replace(['ä','ö','ü','ß'], ['ae','oe','ue','ss'], $Idname);
			if(strcasecmp(trim($name), $Idname) == 0)
				$output = $Id;
			}

		if(isset($output))
			return $output;

		else
			return NULL;

		}

	function getMembers($Id){

		$members = $this->getDetails("15004/$Id");
		
		return $members['9018']['15002']['9003'];

		}

	function getPowerStatus($Id){

		$psid = $this->getDetails("15004/$Id");

		return $psid[ONOFF];

		}

	function poweroff($path){

		$payload = '{ "5850": 0 }';
		$this->action("put", $payload, "15004/$path");

		if($this->getPowerStatus($path) == 0)
			return $this->getName("15004/$path")." wurde ausgeschaltet";

		else
			return $this->getName("15004/$path")." konnte nicht ausgeschaltet werden";

		}

	function poweron($path){

		$payload = '{ "5850": 1 }';
		$this->action("put", $payload, "15004/$path");

		if($this->getPowerStatus($path) == 1)
			return $this->getName("15004/$path")." wurde eingeschaltet";

		else
			return $this->getName("15004/$path")." konnte nicht eingeschaltet werden";

		}

	function setDimmer($path, $dimmer){

		$dim = round(254 * (int)str_replace("%", "", trim($dimmer)) / 100, 0);

		$payload = '{ "'.DIMMER.'": '.$dim.' }';
		$this->action("put", $payload, "15004/$path");

		//if($this->getDimmer($path) == $dim)
			return $this->getName("15004/$path")." wurde auf {$dimmer} gedimmt.";
		/*else
			return $this->getName("15004/$path")." konnte nicht auf {$dimmer} gedimmt werden.";*/

		}

	} // End of class tradfrigroups

?>


defines.php
<?php

//stream identifiers for proc_open
define('STDIN', 0);
define('STDOUT', 1);
define('STDERR', 2);

//IKEA Devices identifiers
define('NAME', 9001);
define('LIGHT', 3311);
define('ONOFF', 5850);					//3311/0/5850 = device on/off
define('DIMMER', 5851);					//3311/0/5851 = device brightness
define('COLORHEX', 5706);				//3311/0/5706
define('COLORX', 5709);					//3311/0/5709 => don't use 5706 and 5709 + 5710 at the same time
define('COLORY', 5710);
define('COLORTEMP', 5711);
define('TRANSITION', 5712);				//Fade Time
define('TYPE', 5750);
define('TYPE_REMOTE_CONTROL', 0);		//5750 = 0 => Default Remote Control
define('TYPE_LIGHT', 2);				//5750 = 2 => All lightning Devices (Driver, Lamps, ...)
define('TYPE_CONTROL_OUTLET', 3);		//5750 = 3 => Control Outlet for power Switching
define('TYPE_MOTION_SENSOR', 4);		//5750 = 4 => Motion Sensor
define('GATEWAY', 15011);
define('GATEWAY_NTP', 9023);
define('GATEWAY_FIRMWARE', 9029);
define('GATEWAY_ALEXA_STATUS', 9093);
define('GATEWAY_GOOGLE_STATUS', 9105);
define('GATEWAY_TIME_UNIX', 9059);
define('GATEWAY_SETUP_TIME', 9069);
define('LAST_SEEN', 9020);

?>


devices.php
<?php

require_once('general.php');

class tradfridevices extends tradfri
	{

	function getDimmer($Id){

		$dimid = $this->getDetails("15001/$Id");

		return $dimid['3311']['0'][DIMMER];

		}

	function getIds(){

		return explode(",", trim(str_replace(['[',']'], "" ,strstr($this->query("15001"), '[65'))));

		}

	function getIdbyName($name){

		$Ids = $this->getIds() ?? [];
		foreach($Ids as $Id){
			$Idname = strtolower(trim($this->getName("15001/$Id")));
			$Idname = str_replace(['ä','ö','ü','ß'], ['ae','oe','ue','ss'], $Idname);
			if(strcasecmp(trim($name), $Idname) == 0)
				$output = $Id;
			}

		if(isset($output))
			return $output;

		else
			return NULL;

		}

	function getTypeId($Id){

		$tid = $this->getDetails("15001/$Id");

		return $tid[TYPE];

		}

	function getPowerStatus($Id){

		$psid = $this->getDetails("15001/$Id");

		switch ($psid[TYPE]) {
			case TYPE_LIGHT:
				return $psid['3311']['0'][ONOFF];
				break;

			case TYPE_CONTROL_OUTLET:
				return $psid['3312']['0'][ONOFF];
				break;
			
			default:
				return NULL;
				break;
			}

		}

	function poweroff($path){

		switch($this->getTypeId($path)){
			case TYPE_LIGHT:
				$payload = '{ "3311": [{ "5850": 0 }] }';
				break;

  			case TYPE_CONTROL_OUTLET:
				$payload = '{ "3312": [{ "5850": 0 }] }';
				break;

			default:
				$payload = NULL;
			}

		if (!is_null($payload)){
			$this->action("put", $payload, "15001/$path");

			if($this->getPowerStatus($path) == 0)
				return $this->getName("15001/$path")." wurde ausgeschaltet";
			else
				return $this->getName("15001/$path")." konnte nicht ausgeschaltet werden";
			}

		else
			return $this->getName("15001/$device")." konnte nicht ausgeschaltet werden, da es keine Lampe ist";

		}

	function poweron($path){

		switch($this->getTypeId($path)){
			case TYPE_LIGHT:
				$payload = '{ "3311": [{ "5850": 1 }] }';
				break;

  			case TYPE_CONTROL_OUTLET:
				$payload = '{ "3312": [{ "5850": 1 }] }';
				break;

			default:
				$payload = NULL;
			}

		if (!is_null($payload)){

			$this->action("put", $payload, "15001/$path");

			if($this->getPowerStatus($path) == 1)
				return $this->getName("15001/$path")." wurde eingeschaltet";
			else
				return $this->getName("15001/$path")." konnte nicht eingeschaltet werden";
			}

		else
			return $this->getName("15001/$device")." konnte nicht eingeschaltet werden, da es keine Lampe ist";

		}

	function setDimmer($path, $dimmer, $transition = NULL){

		if($this->getTypeId($path) == TYPE_LIGHT){

			$dim = round(254 * (int)str_replace("%", "", trim($dimmer)) / 100, 0);

			$payload = is_null($transition) ? '{ "3311": [{ "5851" : '.$dim.' }] }' : '{ "3311": [{ "5851": '.$dim.', "5712": '.$transition.' }] }';
			$this->action("put", $payload, "15001/$path");

			if($this->getDimmer($path) == $dim)
				return $this->getName("15001/$path")." wurde auf {$dimmer} gedimmt";
			else
				return $this->getName("15001/$path")." konnte nicht auf {$dimmer} gedimmt werden";

			}

		else
			return $this->getName("15001/$device")." konnte nicht gedimmt werden, da es keine Lampe ist";

		}

	function statuscontroloutlet(){

		$Ids = $this->getIds();
		sort($Ids, SORT_NUMERIC);
		
		foreach($Ids as $device){
			$details = $this->getDetails("15001/$device");
			if($details[TYPE] == TYPE_CONTROL_OUTLET){
				$output[] = array("id" => $device,"name" => $details[NAME], "type" => $details['3']['1'], "firmware" => $details['3']['3'], "lastseen" => date('H:i:s d.m.Y', $details[LAST_SEEN]), "lastseenunix" => $details[LAST_SEEN], "power" => $details['3312']['0'][ONOFF]);
				}
			}

		return $output;

		}

	function statusremotecontrol(){

		$Ids = $this->getIds();
		sort($Ids, SORT_NUMERIC);

		foreach($Ids as $device){
			$details = $this->getDetails("15001/$device");
			if($details[TYPE] == TYPE_REMOTE_CONTROL || $details[TYPE] == TYPE_MOTION_SENSOR){
				$output[] = array("id" => $device,"name" => $details[NAME], "type" => $details['3']['1'], "battery" => $details['3']['9'], "firmware" => $details['3']['3'], "lastseen" => date('H:i:s d.m.Y', $details[LAST_SEEN]), "lastseenunix" => $details[LAST_SEEN]);
				}
			}

		return $output;

		}

	function statuslamps(){

		$Ids = $this->getIds();
		sort($Ids, SORT_NUMERIC);

		foreach($Ids as $device){
			$details = $this->getDetails("15001/$device");
			if($details[TYPE] == TYPE_LIGHT){

				$colorhex = isset($details[LIGHT]['0'][COLORHEX]) ? $details[LIGHT]['0'][COLORHEX] : NULL;
				$colorx = isset($details[LIGHT]['0'][COLORX]) ? $details[LIGHT]['0'][COLORX] : NULL;
				$colory = isset($details[LIGHT]['0'][COLORY]) ? $details[LIGHT]['0'][COLORY] : NULL;
				$colortemp = isset($details[LIGHT]['0'][COLORTEMP]) ? $details[LIGHT]['0'][COLORTEMP] : NULL;
				$transition = isset($details[LIGHT]['0'][TRANSITION]) ? $details[LIGHT]['0'][TRANSITION] : NULL;

				//Set Dimmer to the INT Value in %
				$details[LIGHT]['0'][DIMMER] = round($details[LIGHT]['0'][DIMMER] * 100 / 255);

				$output[] = array("id" => $device,"name" => $details[NAME], "type" => $details['3']['1'], "power" => $details[LIGHT]['0'][ONOFF], "dimmer" => $details[LIGHT]['0'][DIMMER], "colorhex" => $colorhex, "colorx" => $colorx, "colory" => $colory, "colortemp" => $colortemp, "transition" => $transition, "firmware" => $details['3']['3'], "lastseen" => date('H:i:s d.m.Y', $details[LAST_SEEN]), "lastseenunix" => $details[LAST_SEEN]);

				//Clean up
				unset($colorhex, $colorx, $colory, $colortemp, $transition);

				}
			}

		return $output;

		//Clean up
		unset($Ids, $device, $output);

		}

	}

?>


general.php
<?php

require_once('defines.php');

class tradfri {

	private $gateway;

	function __construct($user, $secret, $gwip){

		$this->gateway['user'] = $user;
		$this->gateway['secretkey'] = $secret;
		$this->gateway['ip'] = $gwip;

		}
		
	function getGatewayData(){

		return $this->gateway;

 		}

	function query($path){

		$cmd = "coap-client -m get -u '{$this->gateway['user']}' -k '{$this->gateway['secretkey']}' 'coaps://{$this->gateway['ip']}:5684/{$path}'";
		$process = proc_open($cmd, [STDOUT => ['pipe', 'w'], STDERR => ['pipe', 'w']], $output);

		//read the outputs
		$stdout = stream_get_contents($output[STDOUT]);
		$stderr = stream_get_contents($output[STDERR]);

		//clean up and properly close our handles
		fclose($output[STDOUT]);
		fclose($output[STDERR]);
		$rc = proc_close($process);

		//$result = json_decode(strstr($stdout,'{"'), true);
		$result = json_encode($stdout, true);

		return $stdout;
		//return $result;

		}

	function action($method, $payload, $path){
		if($payload != "")
			$cmd = "coap-client -m {$method} -u '{$this->gateway['user']}' -k '{$this->gateway['secretkey']}' -e '{$payload}' 'coaps://{$this->gateway['ip']}:5684/{$path}'";
		else
			$cmd = "coap-client -m {$method} -u '{$this->gateway['user']}' -k '{$this->gateway['secretkey']}' 'coaps://{$this->gateway['ip']}:5684/{$path}'";

		return exec($cmd);

		}

	function getDeviceIds(){

		return explode(",", trim(str_replace(['[',']'], "" ,strstr($this->query("15001"), '[65'))));

		}

	function getDetails($path){

		return json_decode(strstr($this->query($path), '{"'), true);

		}

	function getName($path){

		$details = $this->getDetails($path);

		return trim($details[NAME]);

		}

	} //End of Class tradfri

?>
